enhancements/machine-config/mco-flattened-ignition.md

```diff
@@ -0,0 +1,176 @@
+although we need to ensure the approach taken in each case is aligned.
+
+### User Stories
+
+#### Story 1
+
+As a UPI user, I would like to perform network configuration in the same
+way as other customization, e.g via MachineConfig manifests or ignition.
+
+#### Story 2
+
+As a baremetal IPI user, I need to deploy in an environment where network
+configuration is required before access to the controlplane network is possible.
+
+## Design Details
+
+The MCO currently maintains two ignition artifacts, a managed instance of the
+pointer ignition config, stored in a Secret, and the rendered config stored
+in a MachineConfig object (both per role, not per machine).
+
+Both artifacts are accessible via the MCS (although not currently from within
+the cluster ref https://github.com/openshift/machine-config-operator/issues/1690)
+so they may be consumed during deployment.
+
+The proposal is to create a third managed artifact, which is the flattened
+combination of the pointer and rendered ignition configs, so that when
+required it may be consumed instead of requiring runtime download of the rendered
+config via the pointer URL.
+
+Open Questions for discussion:
+  * Should the new artifact be another MachineConfig or a Secret or some other entity?
+  * How do we resolve the issue of CAPBM not being able to access the MCS,
+    can we reliably determine the MCO managed resource and reference it directly?
+  * Does this approach fit with discussions around declarative network
+    configuration (for day-1 and day-2)?
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+## Alternatives
+
+### Native ignition support for early-network config
+
+This was initially proposed via [an ignition issue](https://github.com/coreos/ignition/issues/979)
+where we had some good discussion, but ultimately IIUC the idea of adding a new
+interface for early network configuration to ignition was rejected.
+
+### Implement config flattening in platform-specific repos
+
+The ignition configuration is consumed at two points during deployment for IPI baremetal
+deployments:
+
+  * Via terraform (when the installer deploys the controlplane
+    nodes using [terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic).
+  * Workers are then deployed via the machine-api, using a
+    [baremetal Cluster API Provider](https://github.com/openshift/cluster-api-provider-baremetal/)
+
+The ironic terraform provider is designed to be generic, and thus we would prefer
+not to add OS specific handling of the user-data there, but it would be technically
+possible to do the flattening in some terraform provider, and duplicate that
+logic in the cluster-api-provider-baremetal.
```

Comment:
I experimented with this a little bit using an ignition data source that can create the flattened ignition (https://github.com/openshift-metal3/terraform-provider-openshift).

The problem is that terraform is frustratingly limited when you consider non-cloud cases. I constantly run into issues related to wanting to use terraform to provision infrastructure you'll use later by other terraform resources.  This is one of them: it needs the ignition for the control plane prior to creating the bootstrap.

I had some crazy ideas about throwing the control plane terraform code away, running cluster-baremetal-operator on the bootstrap, and having it provision the control plane. But, it's not a fully thought through idea.

Reply:
@stbenjam thanks for clarifying, I'd forgotten that detail - would you like to propose some alternative text to explain why it's actually not possible, or shall I just rework this to say it's been investigated and isn't possible due to terraform limitations?

Reply:
(this is a tangent)

I can't speak for the installer team, but I think there's general agreement that Terraform is an implementation detail today - we aren't tied to it forever.  And certainly whenever we go to add a kubevirt backend to the installer (yes it makes sense for testing at least) using TF to write Kube objects will be...weird.

I think indeed it would be better if we moved to mostly having TF provision the bootstrap node and *some* associated cloud resources but I know the installer team was concerned about speed.  Not clear to me how valid that is.

Reply:
Ok I'll try to clarify this, I guess the main point is we don't want to do it this way due to duplication and terraform limitations, it will be preferable to have all the ignition artifacts managed in one place (the MCO)

Reply:
I added a comment that this isn't implementable based on previous experience prototyping